File: frontend/src/api/monitoring.ts
import apiFetch from './client'

export interface AgentMetrics {
  [agentId: string]: {
    agent_id: string; agent_name: string
    qps: number; success_rate: number
    latency_p50: number; latency_p95: number; latency_p99: number
    memory_mb: number; cpu_percent: number
    health_score: number; recorded_at: string | null
  }
}

export interface AlertConfig {
  id: string; name: string; description: string
  priority: string; metric_name: string
  operator: string; threshold: number
  duration_seconds: number
  target_type: string; target_agent_id: string | null
  notify_method: string; notify_target: string
  enabled: boolean; created_at: string | null
}

export interface AlertRecord {
  id: string; config_id: string; alert_name: string
  priority: string; agent_id: string
  metric_name: string; current_value: number
  threshold: number; operator: string
  status: string; acknowledged_by: string | null
  acknowledged_at: string | null; resolved_at: string | null
  fired_at: string | null
}

export interface DashboardPanel {
  id: string; title: string; chart_type: string
  metric_names: string[]; agent_ids: string[]
  position_x: number; position_y: number
  width: number; height: number
  config: Record<string, unknown>
  enabled: boolean; created_by: string | null
}

// Metrics
export async function recordMetric(data: Record<string, unknown>): Promise<{ id: string; health_score: number }> {
  const resp = await apiFetch('/api/v1/monitoring/metrics', { method: 'POST', data })
  return resp.data
}

export async function getLatestMetrics(): Promise<AgentMetrics> {
  const resp = await apiFetch('/api/v1/monitoring/metrics/latest', { method: 'GET' })
  return resp.data
}

export async function getMetricHistory(agentId: string, metricNames = 'health_score,qps,latency_p95', hours = 24, interval = 5): Promise<Record<string, Array<{ time: string; value: number }>>> {
  const resp = await apiFetch(`/api/v1/monitoring/metrics/history/${agentId}?metric_names=${encodeURIComponent(metricNames)}&hours=${hours}&interval_minutes=${interval}`, { method: 'GET' })
  return resp.data
}

export async function getAgentRanking(sortBy = 'health_score', limit = 20): Promise<Array<Record<string, unknown>>> {
  const resp = await apiFetch(`/api/v1/monitoring/metrics/ranking?sort_by=${sortBy}&limit=${limit}`, { method: 'GET' })
  return resp.data
}

// Alert configs
export async function createAlertConfig(data: Record<string, unknown>): Promise<AlertConfig> {
  const resp = await apiFetch('/api/v1/monitoring/alert-configs', { method: 'POST', data })
  return resp.data
}

export async function updateAlertConfig(id: string, data: Record<string, unknown>): Promise<AlertConfig> {
  const resp = await apiFetch(`/api/v1/monitoring/alert-configs/${id}`, { method: 'PUT', data })
  return resp.data
}

export async function listAlertConfigs(enabledOnly = false): Promise<AlertConfig[]> {
  const resp = await apiFetch(`/api/v1/monitoring/alert-configs?enabled_only=${enabledOnly}`, { method: 'GET' })
  return resp.data
}

export async function deleteAlertConfig(id: string): Promise<void> {
  await apiFetch(`/api/v1/monitoring/alert-configs/${id}`, { method: 'DELETE' })
}

// Alert records
export async function listAlerts(params: Record<string, string | number | undefined> = {}): Promise<{ data: AlertRecord[]; total: number }> {
  const qs = new URLSearchParams()
  Object.entries(params).forEach(([k, v]) => { if (v != null) qs.set(k, String(v)) })
  const resp = await apiFetch(`/api/v1/monitoring/alerts?${qs}`, { method: 'GET' })
  return resp.data
}

export async function acknowledgeAlert(id: string, userId = ''): Promise<AlertRecord> {
  const resp = await apiFetch(`/api/v1/monitoring/alerts/${id}/acknowledge`, { method: 'POST', data: { user_id: userId } })
  return resp.data
}

export async function resolveAlert(id: string): Promise<AlertRecord> {
  const resp = await apiFetch(`/api/v1/monitoring/alerts/${id}/resolve`, { method: 'POST' })
  return resp.data
}

// Panels
export async function createPanel(data: Record<string, unknown>): Promise<DashboardPanel> {
  const resp = await apiFetch('/api/v1/monitoring/panels', { method: 'POST', data })
  return resp.data
}

export async function updatePanel(id: string, data: Record<string, unknown>): Promise<DashboardPanel> {
  const resp = await apiFetch(`/api/v1/monitoring/panels/${id}`, { method: 'PUT', data })
  return resp.data
}

export async function listPanels(): Promise<DashboardPanel[]> {
  const resp = await apiFetch('/api/v1/monitoring/panels', { method: 'GET' })
  return resp.data
}

export async function deletePanel(id: string): Promise<void> {
  await apiFetch(`/api/v1/monitoring/panels/${id}`, { method: 'DELETE' })
}

File: backend/app/models/monitoring.py
"""
多智能体监控看板模型 — 指标 / 告警 / 面板
"""
import uuid
from datetime import datetime, timezone

from sqlalchemy import Column, String, Text, DateTime, Integer, Float, Boolean
from sqlalchemy.sql import func

from app.db.session import Base


class AgentMetric(Base):
    """Agent 指标时间序列"""
    __tablename__ = "agent_metrics"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    agent_id = Column(String(36), nullable=False, index=True)
    agent_name = Column(String(100))

    # 指标
    qps = Column(Float, default=0.0)
    success_rate = Column(Float, default=100.0)
    latency_p50 = Column(Float, default=0.0)
    latency_p95 = Column(Float, default=0.0)
    latency_p99 = Column(Float, default=0.0)
    memory_mb = Column(Float, default=0.0)
    cpu_percent = Column(Float, default=0.0)

    # 健康评分（0-100）
    health_score = Column(Float, default=100.0)

    # 时间
    recorded_at = Column(DateTime(timezone=True), index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), default=lambda: datetime.now(timezone.utc))


class AlertConfig(Base):
    """告警配置"""
    __tablename__ = "alert_configs"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String(100), nullable=False)
    description = Column(Text)
    priority = Column(String(10), default="P2")  # P0 | P1 | P2 | P3

    # 触发条件
    metric_name = Column(String(50), nullable=False)  # qps | success_rate | latency_p50 | latency_p95 | health_score | memory_mb | cpu_percent
    operator = Column(String(10), nullable=False)  # gt | lt | gte | lte | eq
    threshold = Column(Float, nullable=False)
    duration_seconds = Column(Integer, default=60)  # 持续多久触发

    # 范围
    target_type = Column(String(20), default="all")  # all | specific_agent
    target_agent_id = Column(String(36))

    # 通知
    notify_method = Column(String(50))  # webhook | email | system_message
    notify_target = Column(String(200))

    enabled = Column(Boolean, default=True)

    # 静默管理
    silence_start = Column(String(16), nullable=True)   # "HH:MM" UTC — 静默开始时间
    silence_end = Column(String(16), nullable=True)     # "HH:MM" UTC — 静默结束时间
    silence_days = Column(Text, nullable=True)          # JSON: [0,1,2,3,4,5,6] 星期几(0=周一)
    cooldown_minutes = Column(Integer, default=15)      # 同一告警冷却时间(分钟),避免重复通知

    created_at = Column(DateTime(timezone=True), server_default=func.now(), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), default=lambda: datetime.now(timezone.utc))


class AlertRecord(Base):
    """告警记录"""
    __tablename__ = "alert_records"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    config_id = Column(String(36), index=True)
    alert_name = Column(String(100))
    priority = Column(String(10))

    # 触发详情
    agent_id = Column(String(36), index=True)
    metric_name = Column(String(50))
    current_value = Column(Float)
    threshold = Column(Float)
    operator = Column(String(10))

    # 状态
    status = Column(String(20), default="firing")  # firing | acknowledged | resolved
    acknowledged_by = Column(String(36))
    acknowledged_at = Column(DateTime(timezone=True))
    resolved_at = Column(DateTime(timezone=True))

    fired_at = Column(DateTime(timezone=True), index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), default=lambda: datetime.now(timezone.utc))


class DashboardPanel(Base):
    """自定义看板面板"""
    __tablename__ = "dashboard_panels"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    title = Column(String(100), nullable=False)
    chart_type = Column(String(30))  # line | bar | radar | gauge | number | table
    metric_names = Column(Text)  # JSON: ["qps", "latency_p95"]
    agent_ids = Column(Text)  # JSON: ["agent-1", "agent-2"]

    # 布局
    position_x = Column(Integer, default=0)
    position_y = Column(Integer, default=0)
    width = Column(Integer, default=2)
    height = Column(Integer, default=2)

    # 自定义配置
    config = Column(Text)  # JSON

    enabled = Column(Boolean, default=True)
    created_by = Column(String(36))
    created_at = Column(DateTime(timezone=True), server_default=func.now(), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), onupdate=func.now(), default=lambda: datetime.now(timezone.utc))
